File: java/src/main/java/dev/realmid/sdk/Claims.java
package dev.realmid.sdk;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Verified token payload. Standard JWT fields plus the RealmID-specific
 * extras (azp, tenantId, role). Unknown fields land in {@link #extra}.
 */
public final class Claims {
    private final String issuer;
    private final String subject;
    private final String audience;
    private final long issuedAt;
    private final long notBefore;
    private final long expiry;
    private final String jwtId;
    private final String authorizedParty;
    private final String tenantId;
    private final String role;
    private final long mfaAt;
    private final Map<String, Object> extra;

    /** Backward-compat overload — leaves {@code mfaAt} as 0 (absent). */
    public Claims(
            String issuer,
            String subject,
            String audience,
            long issuedAt,
            long notBefore,
            long expiry,
            String jwtId,
            String authorizedParty,
            String tenantId,
            String role,
            Map<String, Object> extra) {
        this(issuer, subject, audience, issuedAt, notBefore, expiry, jwtId,
                authorizedParty, tenantId, role, 0L, extra);
    }

    public Claims(
            String issuer,
            String subject,
            String audience,
            long issuedAt,
            long notBefore,
            long expiry,
            String jwtId,
            String authorizedParty,
            String tenantId,
            String role,
            long mfaAt,
            Map<String, Object> extra) {
        this.issuer = issuer;
        this.subject = subject;
        this.audience = audience;
        this.issuedAt = issuedAt;
        this.notBefore = notBefore;
        this.expiry = expiry;
        this.jwtId = jwtId;
        this.authorizedParty = authorizedParty;
        this.tenantId = tenantId;
        this.role = role;
        this.mfaAt = mfaAt;
        this.extra = extra == null ? Collections.emptyMap() : Collections.unmodifiableMap(extra);
    }

    public String issuer() { return issuer; }
    public String subject() { return subject; }
    public String audience() { return audience; }
    public long issuedAt() { return issuedAt; }
    public long notBefore() { return notBefore; }
    public long expiry() { return expiry; }
    public String jwtId() { return jwtId; }
    public String authorizedParty() { return authorizedParty; }
    public String tenantId() { return tenantId; }
    public String role() { return role; }

    /**
     * Unix-seconds of the user's most recent successful MFA challenge
     * (SPEC §10.4). 0 means MFA never verified for this session.
     */
    public long mfaAt() { return mfaAt; }

    /** Custom (non-reserved) claims as a read-only map. */
    public Map<String, Object> extra() { return extra; }

    /**
     * Legacy MFA marker check — true when the token carries an
     * {@code amr} array containing {@code "mfa"} or
     * {@code acr == "urn:realmid:mfa"}. Useful as a fallback for tokens
     * minted before {@link #mfaAt()} was introduced. New gates should
     * prefer {@link #mfaAt()} for proof of freshness.
     */
    public boolean hasMfa() {
        Object amr = extra.get("amr");
        if (amr instanceof List<?> list) {
            for (Object x : list) {
                if ("mfa".equals(String.valueOf(x))) return true;
            }
        }
        Object acr = extra.get("acr");
        return "urn:realmid:mfa".equals(acr);
    }
}

File: go/claims.go
package realmid

// Claims is the verified token payload. Standard JWT fields plus the
// RealmID-specific extras (azp, tenant_id, role). Unknown fields land in
// Extra.
type Claims struct {
	Issuer          string         `json:"iss,omitempty"`
	Subject         string         `json:"sub,omitempty"`
	Audience        string         `json:"aud,omitempty"`
	IssuedAt        int64          `json:"iat,omitempty"`
	NotBefore       int64          `json:"nbf,omitempty"`
	Expiry          int64          `json:"exp,omitempty"`
	JWTID           string         `json:"jti,omitempty"`
	AuthorizedParty string         `json:"azp,omitempty"`
	TenantID        string         `json:"tenant_id,omitempty"`
	Role            string         `json:"role,omitempty"`
	AMR             []string       `json:"amr,omitempty"`
	ACR             string         `json:"acr,omitempty"`
	// MFAAt is the unix-seconds timestamp of the user's most recent
	// successful MFA challenge in this session. Zero means absent — the
	// session never completed MFA, or the server hasn't been upgraded
	// yet to emit this claim. SPEC §10.4.
	MFAAt int64          `json:"mfa_at,omitempty"`
	Extra map[string]any `json:"-"`
}

// reservedClaimKeys must stay in sync with the Claims struct fields above.
var reservedClaimKeys = map[string]struct{}{
	"iss": {}, "sub": {}, "aud": {}, "iat": {}, "nbf": {}, "exp": {},
	"jti": {}, "azp": {}, "tenant_id": {}, "role": {}, "amr": {}, "acr": {},
	"mfa_at": {},
}

// HasMFA reports whether the verified claims indicate the user passed an
// MFA challenge — either via amr containing "mfa" or any non-empty acr.
// This is a shape check; for freshness-aware gating, the middleware uses
// the mfa_at claim per SPEC §10.4.
func (c *Claims) HasMFA() bool {
	if c == nil {
		return false
	}
	for _, m := range c.AMR {
		if m == "mfa" {
			return true
		}
	}
	return c.ACR != ""
}
